File: src/main/java/io/github/jsnimda/common/input/GlobalInputHandler.java
package io.github.jsnimda.common.input;

import java.util.ArrayList;
import java.util.List;

import org.lwjgl.glfw.GLFW;

import io.github.jsnimda.common.gui.DebugScreen.DebugInfos;
import io.github.jsnimda.common.input.KeybindSettings.Context;
import io.github.jsnimda.common.input.KeybindSettings.KeyAction;
import net.minecraft.client.MinecraftClient;

public class GlobalInputHandler {

  private static final GlobalInputHandler INSTANCE = new GlobalInputHandler();
  public static GlobalInputHandler getInstance() {
    return INSTANCE;
  }

  public final List<Integer> pressingKeys = new ArrayList<>();
  public List<Integer> beforePressingKeys = new ArrayList<>();
  public int lastKey = -1;
  public int lastAction = -1;

  private GlobalInputHandler() {

  }

  public boolean isActivated(List<Integer> keyCodes, KeybindSettings settings) {
    if (settings.activateOn == KeyAction.PRESS && lastAction == GLFW.GLFW_RELEASE)
      return false;
    if (settings.activateOn == KeyAction.RELEASE && lastAction == GLFW.GLFW_PRESS)
      return false;
    if (settings.context != Context.ANY && MinecraftClient.getInstance().currentScreen != null)
      return false;
    if (settings.context == Context.GUI && MinecraftClient.getInstance().currentScreen == null)
      return false;
    // checked: context, activateOn
    if (keyCodes.isEmpty())
      return false;
    // ref: malilib KeybindMulti.updateIsPressed()
    List<Integer> pressedKeys = lastAction == GLFW.GLFW_PRESS ? pressingKeys : beforePressingKeys;
    if (pressedKeys.size() >= keyCodes.size() && (settings.allowExtraKeys || pressedKeys.size() == keyCodes.size())) {
      if (settings.orderSensitive) {
        for (int i = 0; i < keyCodes.size(); i++) {
          if (keyCodes.get(keyCodes.size() - 1 - i) != pressedKeys.get(pressedKeys.size() - 1 - i)) {
            return false;
          }
        }
        return true;
      } else { // order insensitive
        return keyCodes.contains(lastKey) && pressedKeys.containsAll(keyCodes);
      }
    } else {
      return false;
    }
  }

  public boolean onKeyPress(int key) {
    if (pressingKeys.contains(key)) return false; // should err
    beforePressingKeys = new ArrayList<>(pressingKeys);
    pressingKeys.add(key);
    lastKey = key;
    lastAction = GLFW.GLFW_PRESS;
    return false;
  }

  public boolean onKeyRelease(int key) {
    if (!pressingKeys.contains(key)) return false; // should err
    beforePressingKeys = new ArrayList<>(pressingKeys);
    pressingKeys.remove((Object)key);
    lastKey = key;
    lastAction = GLFW.GLFW_RELEASE;
    return false;
  }


  public boolean onKey(int key, int scanCode, int action, int modifiers) {
    DebugInfos.onKey(key, scanCode, action, modifiers);
    if (action == GLFW.GLFW_PRESS) {
      return onKeyPress(key);
    }
    if (action == GLFW.GLFW_RELEASE) {
      return onKeyRelease(key);
    }
    return false;
  }
  public boolean onMouseButton(int button, int action, int mods) {
    DebugInfos.onMouseButton(button, action, mods);
    if (action == GLFW.GLFW_PRESS) {
      return onKeyPress(button - 100);
    }
    if (action == GLFW.GLFW_RELEASE) {
      return onKeyRelease(button - 100);
    }
    return false;
  }


}

File: src/main/java/io/github/jsnimda/common/input/GlobalInputHandler.kt
package io.github.jsnimda.common.input

import io.github.jsnimda.common.IInputHandler
import io.github.jsnimda.common.gui.debug.DebugInfos
import io.github.jsnimda.common.vanilla.Vanilla
import org.lwjgl.glfw.GLFW.*

object GlobalInputHandler {

  val pressedKeys = mutableSetOf<Int>()
  var previousPressedKeys = pressedKeys.toSet()
    private set
  var lastKey = -1
    private set
  var lastAction = -1 // only GLFW_PRESS or GLFW_RELEASE
    private set

  fun isActivated(keyCodes: List<Int>, settings: KeybindSettings): Boolean {
    if (keyCodes.isEmpty()) return false
    if (!settings.activateOn.isValid(lastAction)) return false
    if (!settings.context.isValid(Vanilla.screen())) return false
    // checked: context, activateOn
    // ref: malilib KeybindMulti.updateIsPressed()
    val validateKeys = if (lastAction == GLFW_PRESS) pressedKeys else previousPressedKeys
    return settings.validates(validateKeys, keyCodes)
  }

  private fun onKey(key: Int, action: Int): Boolean { // action: only GLFW_PRESS or GLFW_RELEASE
    val isPress = action == GLFW_PRESS
    if (isPress == pressedKeys.contains(key)) // (PRESS && contain) || (RELEASE && !contain)
      return false // should err / cancelled by other mod
    previousPressedKeys = pressedKeys.toSet()
    if (isPress)
      pressedKeys.add(key)
    else
      pressedKeys.remove(key)
    lastKey = key
    lastAction = action
    return onInput()
  }

  private fun onInput(): Boolean {
    if (currentAssigningKeybind != null) {
      handleAssignKeybind()
      return true
    }
    registered.forEach { it.onInput(lastKey, lastAction) }
    return false
  }

  // ============
  // Assign keybind
  // ============
  var currentAssigningKeybind: IKeybind? = null
    set(value) {
      pressedFirstKey = false
      ignoreLeftClick = true // left down -> ignore, left up -> set false
      field = value
    }
  private var pressedFirstKey = false
  private var ignoreLeftClick = false // fix forge version while compatible with fabric version

  private fun handleAssignKeybind() {
    val pressedKeys: List<Int> = currentAssigningKeybind
      ?.run { settings.modifierKey.handleKeys(pressedKeys.toList()) }
      ?: pressedKeys.toList()
    if (lastAction == GLFW_PRESS) {
      if (lastKey == KeyCodes.MOUSE_BUTTON_1 && ignoreLeftClick) { // GLFW_MOUSE_BUTTON_1 - 100
        return
      }
      pressedFirstKey = true
      if (lastKey == GLFW_KEY_ESCAPE) { // clear keybind
        currentAssigningKeybind?.keyCodes = listOf()
        currentAssigningKeybind = null
      } else {
        currentAssigningKeybind?.keyCodes = pressedKeys
      }
    } else { // lastAction == GLFW_RELEASE
      if (lastKey == KeyCodes.MOUSE_BUTTON_1) {
        ignoreLeftClick = false
      }
      if (pressedKeys.isEmpty() && pressedFirstKey) {
        currentAssigningKeybind = null // all key released, assignment end
      }
    }
  }

  // ============
  // Vanilla key hook handler
  // ============
  fun onKey(key: Int, scanCode: Int, action: Int, modifiers: Int): Boolean {
    DebugInfos.onKey(key, scanCode, action, modifiers)
    return when (action) {
      GLFW_PRESS, GLFW_RELEASE -> onKey(key, action)
      else -> false
    }
  }

  fun onMouseButton(button: Int, action: Int, mods: Int): Boolean {
    DebugInfos.onMouseButton(button, action, mods)
    return when (action) {
      GLFW_PRESS, GLFW_RELEASE -> onKey(button - 100, action)
      else -> false
    }
  }

  // ============
  // api
  // ============
  private val registered: MutableSet<IInputHandler> = mutableSetOf()

  fun register(inputHandler: IInputHandler): Boolean =
    registered.add(inputHandler)

  fun unregister(inputHandler: IInputHandler): Boolean =
    registered.remove(inputHandler)
}